File: src/geo.js
import * as Cesium from 'cesium';
import { EARTH_RADIUS, BALLOON_MAX_ALT, params } from './config.js';

// ---------------------------------------------------------------------------
// Radio line-of-sight
// ---------------------------------------------------------------------------
export function horizonKm(heightM) {
  return params.horizonRefractionCoeff * Math.sqrt(Math.max(heightM, 0));
}

// The largest possible link range right now: two nodes both at the highest
// possible altitude a balloon can be. Towers are always lower than
// BALLOON_MAX_ALT, so this is a safe (if slightly generous) upper bound for
// ANY pair, balloon-balloon or balloon-tower. Used to size spatial-grid
// neighbor searches so they can't miss a valid pair. Recomputed from the
// live params each call, since horizonRefractionCoeff is UI-adjustable.
export function maxPossibleRangeKm() {
  return 2 * horizonKm(BALLOON_MAX_ALT);
}

// Precompute a node's trig/horizon once per tick so repeated pairwise
// comparisons (each node is checked against ~9 grid cells of neighbors)
// don't redo sin/cos/sqrt from scratch every time. Mutates the node,
// stashing the values under underscore-prefixed fields.
export function precomputeNode(node, heightM) {
  node._latRad = Cesium.Math.toRadians(node.lat);
  node._lonRad = Cesium.Math.toRadians(node.lon);
  node._sinLat = Math.sin(node._latRad);
  node._cosLat = Math.cos(node._latRad);
  node._horizonKm = horizonKm(heightM);
}

// Great-circle distance using each node's precomputed trig (see precomputeNode).
export function greatCircleDistanceKmPrecomputed(a, b) {
  const R = EARTH_RADIUS / 1000;
  const dPhi = b._latRad - a._latRad;
  const dLambda = b._lonRad - a._lonRad;
  const sinDPhi = Math.sin(dPhi / 2);
  const sinDLambda = Math.sin(dLambda / 2);
  const h = sinDPhi * sinDPhi + a._cosLat * b._cosLat * sinDLambda * sinDLambda;
  return 2 * R * Math.asin(Math.sqrt(h));
}

export function inRadioRangePrecomputed(a, b) {
  const maxRange = a._horizonKm + b._horizonKm;
  return greatCircleDistanceKmPrecomputed(a, b) <= maxRange;
}

// ---------------------------------------------------------------------------
// Uniform random point on a sphere (avoids clustering near the poles that a
// naive uniform-lat sample would produce).
// ---------------------------------------------------------------------------
export function randomGlobalPosition() {
  const lon = Math.random() * 360 - 180;
  const lat = Cesium.Math.toDegrees(Math.asin(Math.random() * 2 - 1));
  return { lon, lat };
}

export function lerpColor(c1, c2, t) {
  const clamped = Cesium.Math.clamp(t, 0, 1);
  return [
    Math.round(c1[0] + (c2[0] - c1[0]) * clamped),
    Math.round(c1[1] + (c2[1] - c1[1]) * clamped),
    Math.round(c1[2] + (c2[2] - c1[2]) * clamped),
  ];
}

File: sim-server/src/geo.rs
// Ported from src/geo.js. Radio line-of-sight + great-circle
// distance, using each node's precomputed trig so repeated pairwise checks
// don't redo sin/cos/sqrt from scratch every time.

use crate::config::EARTH_RADIUS_M;
use rand::Rng;

pub fn to_radians(deg: f64) -> f64 {
    deg.to_radians()
}

pub fn horizon_km(height_m: f64, horizon_refraction_coeff: f64) -> f64 {
    horizon_refraction_coeff * height_m.max(0.0).sqrt()
}

#[derive(Clone, Copy, Debug, Default)]
pub struct Precomputed {
    pub lat_rad: f64,
    pub lon_rad: f64,
    pub sin_lat: f64,
    pub cos_lat: f64,
    pub horizon_km: f64,
}

pub fn precompute(lon: f64, lat: f64, height_m: f64, horizon_refraction_coeff: f64) -> Precomputed {
    let lat_rad = to_radians(lat);
    let lon_rad = to_radians(lon);
    Precomputed {
        lat_rad,
        lon_rad,
        sin_lat: lat_rad.sin(),
        cos_lat: lat_rad.cos(),
        horizon_km: horizon_km(height_m, horizon_refraction_coeff),
    }
}

pub fn great_circle_distance_km_precomputed(a: &Precomputed, b: &Precomputed) -> f64 {
    let r = EARTH_RADIUS_M / 1000.0;
    let d_phi = b.lat_rad - a.lat_rad;
    let d_lambda = b.lon_rad - a.lon_rad;
    let sin_d_phi = (d_phi / 2.0).sin();
    let sin_d_lambda = (d_lambda / 2.0).sin();
    let h = sin_d_phi * sin_d_phi + a.cos_lat * b.cos_lat * sin_d_lambda * sin_d_lambda;
    2.0 * r * h.sqrt().asin()
}

pub fn in_radio_range_precomputed(a: &Precomputed, b: &Precomputed) -> bool {
    let max_range = a.horizon_km + b.horizon_km;
    great_circle_distance_km_precomputed(a, b) <= max_range
}

/// Uniform random point on a sphere (avoids clustering near the poles that a
/// naive uniform-lat sample would produce).
pub fn random_global_position(rng: &mut impl Rng) -> (f64, f64) {
    let lon = rng.gen::<f64>() * 360.0 - 180.0;
    let lat = (rng.gen::<f64>() * 2.0 - 1.0).asin().to_degrees();
    (lon, lat)
}
